### PR TITLE
Refs Task#389625 & #391920 Send message to chatwork and handle permission error

### PR DESCRIPTION
**Pull request by ghost**
## Related Tickets

- [#389625](https://dev.sun-asterisk.com/issues/389625)
- [#391920](https://dev.sun-asterisk.com/issues/391920)
- [#391923](https://dev.sun-asterisk.com/issues/391923)
## What's this PR do ?

- [x] Implement function send message to chatwork
- [x] Handle error permission denined
- [x] Handle error limit request
## Library
*(List package, library third party add new include version)*

- N/A

## Performance

- [ ] Resolved n + 1 query
- [ ] Run explain query already
- [ ] Time run rake task : 1000 ms

## Checklist

- [x] It was tested in local success?
- [x] Fill link PR into ticket and the opposite
- [x] Note purpose/reason, solution, scope of influence into ticket

## ENV note

```
```

## Notes
*(Other notes)*

### CODE REVIEW

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>failure</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr></table>

File: app/Jobs/SendMessageToChatwork.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+{
+    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;
+
+    protected $bot;
+    protected $room_id;
+    protected $messages;
+    /**
+     * Create a new job instance.
+     *
+     * @return void
```
Comment (framgiaciv3) [inline app/Jobs/SendMessageToChatwork.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Jobs/SendMessageToChatwork.php</br>**Line**: 27</br>Variable "room_id" is not in valid camel caps format

File: app/Jobs/SendMessageToChatwork.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    public function __construct(Bot $bot, $room_id, $messages)
+    {
```
Comment (framgiaciv3) [inline app/Jobs/SendMessageToChatwork.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Jobs/SendMessageToChatwork.php</br>**Line**: 30</br>Variable "room_id" is not in valid camel caps format

File: app/Services/ForwardChatworkService.php [hunk 3]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
 
         return '';
     }
+
+    public function sendMessages($messages, $payload_history_id)
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 134</br>Variable "payload_history_id" is not in valid camel caps format

File: app/Services/ForwardChatworkService.php [hunk 4]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
...
+    {
+        $bot = $this->webhook->bot;
+        $room_id = $this->webhook->room_id;
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 137</br>Variable "room_id" is not in valid camel caps format

File: app/Services/ForwardChatworkService.php [hunk 5]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
...
+        $chatwork = Chatwork::withAPIToken($bot->bot_key);
+
+        foreach($messages as $key=>$message) {
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.ControlStructures.ControlSignature.SpaceAfterKeyword</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 140</br>Expected 1 space(s) after FOREACH keyword; 0 found

Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.ControlStructures.ForEachLoopDeclaration.NoSpaceBeforeArrow</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 140</br>Expected 1 space before "=>"; 0 found

Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.ControlStructures.ForEachLoopDeclaration.NoSpaceAfterArrow</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 140</br>Expected 1 space after "=>"; 0 found

File: app/Services/ForwardChatworkService.php [hunk 6]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
...
+            try {
+                $chatwork->room($room_id)->messages()->create($message);
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 142</br>Variable "room_id" is not in valid camel caps format

File: app/Services/ForwardChatworkService.php [hunk 7]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
...
+                array_splice($messages, $key, 1);
+                // save success message_history
+                $this->saveMessageHistory($payload_history_id, $message, MessageHistoryStatus::SUCCESS);
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 145</br>Variable "payload_history_id" is not in valid camel caps format

File: app/Services/ForwardChatworkService.php [hunk 8]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
...
+            } catch (APIException $error) {
+                switch ($error->getStatus()) {
+                    case 403:
+                        // permission denined
+                        $this->saveMessageHistory($payload_history_id, $message, MessageHistoryStatus::FAILED, 'Permission');
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 150</br>Variable "payload_history_id" is not in valid camel caps format

Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Generic.Files.LineLength.TooLong</br>**Severity**: warning</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 150</br>Line exceeds 120 characters; contains 125 characters

File: app/Services/ForwardChatworkService.php [hunk 9]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
...
+                        break 2;
+                    case 429:
+                        // limit request
+                        // dd('asdasd');
+                        SendMessageToChatwork::dispatch($bot, $room_id, $messages)->delay(now()->addSeconds(5));
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 155</br>Variable "room_id" is not in valid camel caps format

File: app/Services/ForwardChatworkService.php [hunk 10]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
...
+                        break 2;
+                    case 401:
+                        // authorized
+                        $this->saveMessageHistory($payload_history_id, $message, MessageHistoryStatus::FAILED, 'Unauthorized');
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 159</br>Variable "payload_history_id" is not in valid camel caps format

Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Generic.Files.LineLength.TooLong</br>**Severity**: warning</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 159</br>Line exceeds 120 characters; contains 127 characters

File: app/Services/ForwardChatworkService.php [hunk 11]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
...
+                // permission -> save failed message_history, kill all payload 403
+
+                // limited request - retry after 5 minutes/ 429
+
+                // Request timeout -> retry 3 times
+            }
+        }
+    }
+
+    public function saveMessageHistory($payload_history_id, $message, $status, $log = '')
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 172</br>Variable "payload_history_id" is not in valid camel caps format

File: app/Services/ForwardChatworkService.php [hunk 12]
```diff
@@ -117,4 +130,52 @@ function ($match) use ($params, &$isMatching) {
...
+    {
+        $this->messageHistoryRepository->create([
+            'payload_history_id' => $payload_history_id,
```
Comment (framgiaciv3) [inline app/Services/ForwardChatworkService.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Framgia.Variables.ValidVariableName.NotCamelCaps</br>**Severity**: error</br>**File**: app/Services/ForwardChatworkService.php</br>**Line**: 175</br>Variable "payload_history_id" is not in valid camel caps format

**Review by nguyenthinhhn — APPROVED** [review]

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr></table>

File: app/Http/Controllers/ForwardChatworkController.php [hunk 13]
```diff
@@ -7,26 +7,37 @@
...
+            );
+
             $forwardChatworkService->call();
-        } else {
-            return null;
+
+            return response()->json('Excuted successfully');
         }
+
+        return response()->json('Webhook not found. Please try agian');
```
Comment (ghost) [inline app/Http/Controllers/ForwardChatworkController.php]: sai chính tả nè again

File: app/Services/ForwardChatworkService.php [hunk 14]
```diff
@@ -117,4 +131,45 @@ function ($match) use ($params, &$isMatching) {
...
+                $this->saveMessageHistory($message, MessageHistoryStatus::SUCCESS);
+            } catch (APIException $error) {
+                switch ($error->getStatus()) {
+                    case 403:
+                        // permission denined
+                        $this->saveMessageHistory($message, MessageHistoryStatus::FAILED, 'Permission');
+                        break 2;
+                    case 429:
+                        // limit request
+                        SendMessageToChatwork::dispatch($bot, $roomId, $messages)->delay(now()->addSeconds(5));
```
Comment (ghost) [inline app/Services/ForwardChatworkService.php]: 5 phút chứ

**Review by nguyenthinhhn — APPROVED** [review]

File: app/Services/ForwardChatworkService.php (hunk 14 above)
Comment (ghost) [reply app/Services/ForwardChatworkService.php]: oke chú, case limit này đang để vậy để test thử á.
chú review case permiss vs send message trước đi nhá. tạm thời bỏ qua mấy case khác đi. pull sau làm thêm

**Review by ghost — APPROVED** [review]
approved đợt 1 :v

File: database/migrations/2019_12_05_064317_create_jobs_table.php [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+class CreateJobsTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('jobs', function (Blueprint $table) {
```
Comment (tamtt-1031) [inline database/migrations/2019_12_05_064317_create_jobs_table.php:16]: vẽ vào design DB luôn nghe. 
https://www.draw.io/#G1ocUzhuWosTYd-jjeXUgXnq2sPXARYZHn

File: app/Jobs/SendMessageToChatwork.php [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        foreach ($this->$messages as $message) {
+            try {
+                $this->chatwork->room($this->roomId)->messages()->create($message);
+                // save success message_history
+                // $this->saveMessageHistory($payload_history_id, $message, MessageHistoryStatus::SUCCESS);
+            } catch (APIException $error) {
+                MessageHistory::create([
+                    'payload_history_id' => 1,
+                    'message_content' => 1,
+                    'status' => 0,
```
Comment (tamtt-1031) [inline app/Jobs/SendMessageToChatwork.php]: Cần định nghĩa enum phân biệt các loại fail do j ko nhỉ?

Comment (ghost) [reply app/Jobs/SendMessageToChatwork.php]: e update r a :D

File: database/migrations/2019_12_05_064317_create_jobs_table.php (hunk 15 above)
Comment (ghost) [reply database/migrations/2019_12_05_064317_create_jobs_table.php:16]: oke a

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>failure</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr></table>

File: app/Services/ForwardChatworkService.php [hunk 17]
```diff
@@ -33,10 +40,15 @@ public function call()
...
             foreach ($conditions as $condition) {
                 // 'return $payloads->fieldName == "value";'
-                $expression = 'return ' . $condition->field . ' '
-                                        . $condition->operator . ' "'
-                                        . $condition->value . '";';
-                if (!eval($expression)) {
+                try {
+                    $expression = 'return ' . $condition->field . ' '
+                                            . $condition->operator . ' "'
+                                            . $condition->value . '";';
```
Comment (tamtt-1031) [inline app/Services/ForwardChatworkService.php]: `. ' " ' . $condition->value . '";';`
đưa thằng " ở trên xuống dưới nó dễ hiểu là `"value";` e

**Comment by framgiaciv3** [comment]
No data coverage!

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.11</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.13</td><td>success</td><td></td></tr><tr><td>yarn_dev._start_run_.9</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.17</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.19</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>yarn_install._start_run_.7</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.15</td><td>success</td><td></td></tr></table>

**Review by ghost — APPROVED** [review]

**Review by quanghung97 — APPROVED** [review]
OK

**Comment by phucqd-1305** [comment]
OK

**Pull request merged**